File: freight-triage-polyglot-ledger-recovery/environment/native/src/hash/jenkins_oaat.cpp
#include "freight/hashes.h"

namespace freight {

// jenkins_oaat over raw bytes.
uint64_t hash_jenkins_oaat(const std::string& data) {
  uint32_t state = 0u;
  for (size_t i = 0; i < data.size(); ++i) {
    state += static_cast<uint8_t>(data[i]);
    state += state << 10;
    state ^= state >> 6;
  }
  state += state << 3;
  state ^= state >> 11;
  state += state << 15;
  return static_cast<uint64_t>(state);
}

}  // namespace freight

File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/hash/JenkinsOaat.java
package com.freight.hash;

/** jenkins_oaat over raw bytes. */
public final class JenkinsOaat implements HashAlgorithm {

    @Override
    public String name() {
        return "jenkins_oaat";
    }

    @Override
    public long apply(byte[] data) {
        long state = 0L;
        for (int i = 0; i < data.length; i++) {
            state = (state + (data[i] & 0xFF)) & 0xFFFFFFFFL;
            state = (state + (state << 10)) & 0xFFFFFFFFL;
            state ^= (state >>> 6);
        }
        state = (state + (state << 3)) & 0xFFFFFFFFL;
        state ^= (state >>> 11);
        state = (state + (state << 15)) & 0xFFFFFFFFL;
        return state;
    }
}
